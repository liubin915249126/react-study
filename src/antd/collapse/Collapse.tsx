import * as React from 'react';
import RcCollapse from 'rc-collapse';
import classNames from 'classnames';
import RightOutlined from '@ant-design/icons/RightOutlined';

import CollapsePanel from './CollapsePanel';
import { ConfigContext } from '../config-provider';
import animation from './openAnimation';
import { cloneElement } from '../_util/reactNode';

export type ExpandIconPosition = 'left' | 'right' | undefined;

export interface CollapseProps {
  activeKey?: Array<string | number> | string | number;
  defaultActiveKey?: Array<string | number> | string | number;
  /** 手风琴效果 */
  accordion?: boolean;
  destroyInactivePanel?: boolean;
  onChange?: (key: string | string[]) => void;
  style?: React.CSSProperties;
  className?: string;
  bordered?: boolean;
  prefixCls?: string;
  expandIcon?: (panelProps: PanelProps) => React.ReactNode;
  expandIconPosition?: ExpandIconPosition;
  ghost?: boolean;
}

interface PanelProps {
  isActive?: boolean;
  header?: React.ReactNode;
  className?: string;
  style?: React.CSSProperties;
  showArrow?: boolean;
  forceRender?: boolean;
  disabled?: boolean;
  extra?: React.ReactNode;
}

interface CollapseInterface extends React.FC<CollapseProps> {
  Panel: typeof CollapsePanel;
}

const Collapse: CollapseInterface = props => {
  const { getPrefixCls, direction } = React.useContext(ConfigContext);
  const { prefixCls: customizePrefixCls, className = '', bordered, ghost } = props;
  const prefixCls = getPrefixCls('collapse', customizePrefixCls);

  const getIconPosition = () => {
    const { expandIconPosition } = props;
    if (expandIconPosition !== undefined) {
      return expandIconPosition;
    }
    return direction === 'rtl' ? 'right' : 'left';
  };

  const renderExpandIcon = (panelProps: PanelProps = {}) => {
    const { expandIcon } = props;
    const icon = (expandIcon ? (
      expandIcon(panelProps)
    ) : (
      <RightOutlined rotate={panelProps.isActive ? 90 : undefined} />
    )) as React.ReactNode;

    return cloneElement(icon, () => ({
      className: classNames((icon as any).props.className, `${prefixCls}-arrow`),
    }));
  };

  const iconPosition = getIconPosition();
  const collapseClassName = classNames(
    {
      [`${prefixCls}-borderless`]: !bordered,
      [`${prefixCls}-icon-position-${iconPosition}`]: true,
      [`${prefixCls}-rtl`]: direction === 'rtl',
      [`${prefixCls}-ghost`]: !!ghost,
    },
    className,
  );
  const openAnimation = { ...animation, appear() {} };

  return (
    <RcCollapse
      openAnimation={openAnimation}
      {...props}
      expandIcon={(panelProps: PanelProps) => renderExpandIcon(panelProps)}
      prefixCls={prefixCls}
      className={collapseClassName}
    />
  );
};

Collapse.Panel = CollapsePanel;

Collapse.defaultProps = {
  bordered: true,
};

export default Collapse;
