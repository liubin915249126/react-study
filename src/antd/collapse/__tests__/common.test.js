import mountTest from '../../../tests/shared/mountTest';
import rtlTest from '../../../tests/shared/rtlTest';
import Collapse from '..';
import openAnimation from '../openAnimation';

describe('Collapse', () => {
  mountTest(Collapse);
  rtlTest(Collapse);
});

describe('openAnimation', () => {
  it('should support openAnimation', () => {
    const done = jest.fn();
    const domNode = document.createElement('div');
    expect(typeof openAnimation.enter).toBe('function');
    expect(typeof openAnimation.leave).toBe('function');
    expect(typeof openAnimation.appear).toBe('function');
    const appear = openAnimation.appear(domNode, done);
    const enter = openAnimation.enter(domNode, done);
    const leave = openAnimation.leave(domNode, done);
    expect(typeof appear.stop).toBe('function');
    expect(typeof enter.stop).toBe('function');
    expect(typeof leave.stop).toBe('function');
    expect(done).toHaveBeenCalled();
  });
});
