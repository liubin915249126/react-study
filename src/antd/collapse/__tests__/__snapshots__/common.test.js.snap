// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collapse rtl render component should be rendered correctly in RTL direction 1`] = `
<div
  class="ant-collapse ant-collapse-icon-position-right ant-collapse-rtl"
/>
`;
